hphp/runtime/vm/jit/vasm-reuseimm.cpp

```diff
@@ -0,0 +1,212 @@
+   | HipHop for PHP                                                       |
+   +----------------------------------------------------------------------+
+   | Copyright (c) 2010-present Facebook, Inc. (http://www.facebook.com)  |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+*/
+
+
+#include "hphp/runtime/vm/jit/vasm.h"
+
+#include "hphp/runtime/vm/jit/vasm-gen.h"
+#include "hphp/runtime/vm/jit/vasm-instr.h"
+#include "hphp/runtime/vm/jit/vasm-print.h"
+#include "hphp/runtime/vm/jit/vasm-reg.h"
+#include "hphp/runtime/vm/jit/vasm-unit.h"
+#include "hphp/runtime/vm/jit/vasm-util.h"
+
+#include "hphp/util/arch.h"
+#include "hphp/util/assertions.h"
+
+#include <cstdlib>
+
+TRACE_SET_MOD(vasm);
+
+namespace HPHP { namespace jit {
+
+namespace {
+
+// track ldimmq values
+struct ImmState {
+  ImmState() : val{0}, base{} {}
+  ImmState(Immed64 a, Vreg b) : val{a}, base{b} {}
+
+  Immed64 val;
+  Vreg base;
+};
+
+struct ImmStateAry {
+  void clear();
+  ImmState& operator[](size_t i) { return immState_ary[i % size]; }
+
+  int size;
+  enum { max_history = 16};
+  ImmState immState_ary[max_history];
+};
+
+void ImmStateAry::clear() {
+  for (int i=0; i < size; i++) {
+    immState_ary[i].base = Vreg{};
+  }
+}
+
+struct Env {
+  Vunit& unit;
+  ImmStateAry& immStateAry;
+};
+
+bool isMultiword(int64_t imm) {
+  switch (arch()) {
+  case Arch::X64:
+  case Arch::PPC64:
+    break;
+  case Arch::ARM:
+    uint64_t val = std::abs(imm);
+    if (val > (1 << 16)) return true;
+    break;
+  }
+  return false;
+}
+
+// candidate around +/- uimm12
+bool reuseCandidate(Env& env, int64_t p, Vreg& reg, int& delta) {
+  delta = 0;
+  for (int i = 0; i < env.immStateAry.size; i++) {
+    ImmState elem = env.immStateAry[i];
+    if (!elem.base.isValid()) continue;
+    int64_t q = elem.val.q();
+    if ((p >= q) && (p < (q + 4095))) {
+      delta = safe_cast<int>(p - q);
+      reg = elem.base;
+      return true;
+    }
+    if ((p < q) && (q < (p + 4095))) {
+      delta = -safe_cast<int>(q - p);
+      reg = elem.base;
+      return true;
+    }
+  }
+  return false;
+}
+
+template <typename Inst>
+void reuseImmq(Env& env, const Inst& inst, Vlabel b, size_t i) { 
+  env.immStateAry[i] = ImmState{};
+  return;
+}
+
+// purge table across any instruction with RegSet or control flow change
```

Comment:
`syncpoint{}` is also a marker instruction, so it should be fine.  As for calls—the register allocator will ensure that any Vregs that are live across a call will be preserved, either by being in a callee-saved register, or by spilling it to the stack.  If you don't special-case calls, correctness won't be affected, but since you're trying to avoid added register pressure, I think it makes sense to handle them in this way.

I would just use `isBlockEnd() || /* is one of the call instructions */` in that case.